hpx/lcos/detail/async_implementations.hpp

```diff
@@ -96,16 +132,51 @@ namespace hpx { namespace detail
         static lcos::future<Result>
         call(naming::id_type const& id, naming::address && addr, Ts &&... vs)
         {
-            future<Result> f;
+            typedef typename hpx::traits::extract_action<Action>::type
+                action_type;
+
+            try
             {
-                handle_managed_target<Result> hmt(id, f);
-                lcos::packaged_action<Action, Result> p;
+                auto&& result = action_type::execute_function(
+                    addr.address_, addr.type_, std::forward<Ts>(vs)...);
 
-                f = p.get_future();
-                p.apply(std::move(addr), hmt.get_id(), std::forward<Ts>(vs)...);
-                f.wait();
+                typedef typename util::decay<decltype(result)>::type naked_type;
+                typedef traits::get_remote_result<Result, naked_type>
+                    get_remote_result_type;
+
+                future<Result> f = make_ready_future(
+                    get_remote_result_type::call(std::move(result)));
+
+                return keep_alive(std::move(f), id);
```

Comment:
I don't think we need to keep the id alive here. The future is already done and the handler will be called immediately having no effect at all.

Reply:
I added this to maintain semantic compatibility with the old code. A user may rely on the id being kept alive until the value was extracted from the future.

Reply:
Hmm, in which scenario would that be the case? Originally, the keep alive was there to ensure that the component stays alive until the result is delivered. In the case where have a local synchronous invocation, this holds true. In addition, the handler is called immediately with the id_type refcounter being decremented.

Reply:
Ok, I concur.